### PR TITLE
[WIP] Adding mixup to timm_image model

### PR DESCRIPTION
**Pull request by Linuxdex**
*Issue #, if available:*

*Description of changes:*
Add mixup with a single function. More mixup mode will be added to expand mixup as a class.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1730-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1730/1/index.html

File: text/src/autogluon/text/automm/data/process_image.py [hunk 1]
```diff
@@ -250,18 +251,40 @@ def construct_processor(
...
+            elif trans_mode == "affine":
+                if args is not None:
+                    processor.append(transforms.RandomAffine(*args))
+                else:
+                    processor.append(transforms.RandomAffine(15, translate=(0.1, 0.1), scale=(0.9, 1.1)))
+            elif trans_mode == "randaug":
+                if args is not None:
+                    processor.append(RandAugment(*args))
+                else:
+                    processor.append(RandAugment(2, 9))
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/data/process_image.py:285]: You may need to first merge `upstream/master`

**Review by bryanyzhu — APPROVED** [review]
The mixup implementation looks correct to me.

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 2]
```diff
@@ -462,3 +464,16 @@ def apply_layerwise_lr_decay(
         parameter_group_names[group_name]["params"].append(name)
 
     return list(parameter_group_vars.values())
+
+def mixup(
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:468]: Why don't we use timm's mixup class? It supports more features, e.g., cutmix. https://github.com/rwightman/pytorch-image-models/blob/20a1fa63f8ea999dab29d927d5e1866ed3b67348/timm/data/mixup.py#L90

File: text/src/autogluon/text/automm/optimization/lit_module.py [hunk 3]
```diff
@@ -44,6 +45,7 @@ def __init__(
             custom_metric_func: Callable = None,
             test_metric: Optional[torchmetrics.Metric] = None,
             efficient_finetune: Optional[str] = None,
+            mixup_alpha: Optional[float] = 0,
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/lit_module.py:48]: We need to pass a mixup_fn here, which should be initialized in `AutoMMPredictor.fit()`. Basically, `AutoMMPredictor.fit()` is where we make preparations for training. `mixup_alpha ` is a data related hyper-parameter, which is not good design to be the input of lit_module.

Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/lit_module.py:48]: Refer to https://github.com/rwightman/pytorch-image-models/blob/master/train.py#L634

File: text/src/autogluon/text/automm/optimization/lit_module.py [hunk 4]
```diff
@@ -174,8 +177,16 @@ def training_step(self, batch, batch_idx):
...
+        if self.mixup_alpha > 0:
+            image = batch['timm_image_image']
+            label = batch[self.model.label_key]
+            mix_images, target_a, target_b, lam = mixup(image, label, alpha=self.mixup_alpha)
+            batch['timm_image_image'] = mix_images
+            output, loss = self._shared_step(batch)
+            self.log("train_loss", loss)
+        else:
+            output, loss = self._shared_step(batch)
+            self.log("train_loss", loss)
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/lit_module.py:189]: We need to have something like this: https://github.com/rwightman/pytorch-image-models/blob/master/train.py#L696-L697 Always have good API design in our mind when adding new stuff.

File: text/tests/unittests/automm/test_data_process_image.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            "model.timm_image.val_transform_types": ["resize_shorter_side","center_crop"]
+        },
+
+        {
+            "model.timm_image.train_transform_types": ["resize_shorter_side","center_crop","affine","color_jitter","randaug"],
+            "model.timm_image.val_transform_types": ["resize_shorter_side","center_crop"]
+        },
+
+        {
+            "model.timm_image.train_transform_types": ["resize_shorter_side","center_crop","affine(15, (0.1, 0.1), (0.9, 1.1))"],
```
Comment (zhiqiangdon) [inline text/tests/unittests/automm/test_data_process_image.py:41]: The hyper-parameters' meanings are unclear if using tuple. I think we need to support dictionary, which is more informative. For example, we can do ast.literal_eval("{'degrees': 15, 'translate': (0.1, 0.1), 'scale': (0.9, 1.1)}").

File: text/tests/unittests/automm/test_data_process_image.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+            "model.timm_image.val_transform_types": ["resize_shorter_side","center_crop"]
+        },
+    ]
+)
+
+def test_data_process_image(augmentations):
+    dataset = ALL_DATASETS["petfinder"]()
+    image = PIL.Image.open(dataset.train_df.Images[0]).convert("RGB")
+
+    data_processors=ImageProcessor(
```
Comment (zhiqiangdon) [inline text/tests/unittests/automm/test_data_process_image.py:61]: `data_processors ` -> `image_processor`

File: text/tests/unittests/automm/test_data_process_image.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        image_column_names=['train_df.Images'],
+        prefix="timm_image",
+        train_transform_types=augmentations["model.timm_image.train_transform_types"],
+        val_transform_types=augmentations["model.timm_image.val_transform_types"],
+        size=224,
+        norm_type="imagenet",
+    )
+
+    transimage = data_processors.train_processor(image)
```
Comment (zhiqiangdon) [inline text/tests/unittests/automm/test_data_process_image.py:70]: `transimage ` -> `transformed_image`

File: text/tests/unittests/automm/test_data_process_image.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        },
+    ]
+)
+
+def test_data_process_image(augmentations):
+    dataset = ALL_DATASETS["petfinder"]()
+    image = PIL.Image.open(dataset.train_df.Images[0]).convert("RGB")
+
+    data_processors=ImageProcessor(
+        image_column_names=['train_df.Images'],
```
Comment (zhiqiangdon) [inline text/tests/unittests/automm/test_data_process_image.py:62]: This column name looks weird. You can check what's image column name in this dataset.

File: text/src/autogluon/text/automm/data/process_image.py (hunk 1 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/data/process_image.py:285]: Need to rebase against the master branch.

**Comment by szha** [comment]
Job PR-1730-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1730/2/index.html

**Comment by szha** [comment]
Job PR-1730-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1730/3/index.html

**Comment by szha** [comment]
Job PR-1730-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1730/6/index.html

**Comment by sxjscience** [comment]
We may later have a tutorial to explain what types of data augmentations we adopted in AutoMM.

File: text/tests/unittests/automm/test_data_process_image.py [hunk 9]
```diff
@@ -42,11 +42,21 @@
...
+
         {
             "model.timm_image.train_transform_types": ["resize_shorter_side","center_crop","color_jitter(0.2, 0.1, 0.1)"],
             "model.timm_image.val_transform_types": ["resize_shorter_side","center_crop"]
         },
 
+        {
+            "model.timm_image.train_transform_types": ["resize_shorter_side","center_crop","color_jitter({'brightness': 0.2, 'contrast': 0.1, 'saturation': 0.1})"],
+            "model.timm_image.val_transform_types": ["resize_shorter_side","center_crop"]
+        },
```
[the remark this answers is not in the document]
Comment (sxjscience) [reply text/tests/unittests/automm/test_data_process_image.py:58]: Mixup + cutmix

**Comment by szha** [comment]
Job PR-1730-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1730/7/index.html

File: text/src/autogluon/text/automm/data/process_image.py [hunk 10]
```diff
@@ -269,17 +273,23 @@ def construct_processor(
...
-                    processor.append(transforms.RandomAffine(15, translate=(0.1, 0.1), scale=(0.9, 1.1)))
+                    processor.append(transforms.RandomAffine(degrees=15, translate=(0.1, 0.1), scale=(0.9, 1.1)))
             elif trans_mode == "randaug":
-                if args is not None:
+                if kargs is not None:
+                    processor.append(RandAugment(**kargs))
+                elif args is not None:
                     processor.append(RandAugment(*args))
                 else:
                     processor.append(RandAugment(2, 9))
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/data/process_image.py:295]: The if-else patterns of `args` and `kwargs` repeat for each augmentation type. Consider make this pattern a separate util function. The default values can also be handled by the `kwargs` choice. For example, if we detect augmentation type `affine`, we can create a default `kwargs` {"degree"=15, "translate"=(0.1, 0.1), "scale"=(0.9, 1.1)} for it.

File: text/src/autogluon/text/automm/optimization/lit_module.py [hunk 11]
```diff
@@ -174,7 +178,12 @@ def training_step(self, batch, batch_idx):
         -------
         Average loss of the mini-batch data.
         """
-        output, loss = self._shared_step(batch)
+        if self.mixup_fn is not None and (self.hparams.mixup_off_epoch is None or self.current_epoch < self.hparams.mixup_off_epoch):
+            batch, target_a, target_b, lam = self.mixup_fn(batch, "timm_image_image", self.model.label_key)
+            output, _ = self._shared_step(batch)
+            loss = self._compute_loss(output=output, label=target_a) * lam + self._compute_loss(output=output,label=target_b) * (1. - lam)
+        else:
+            output, loss = self._shared_step(batch)
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/lit_module.py:186]: This design has limitations since it uses a fixed model prefix `timm_image_image`. Users may customize this key and there are maybe multiple image backbones, e.g., `timm_image_0`, `timm_image_1`. Multimodal models, e.g., CLIP, also has its image backbone. 

To make mixup scalable to these situations, I think we should move mixup into the model definitions, e.g, timm_image.py and clip.py. We can let the model's output also include the label information, e.g., {"label": {"label_a": label_a, "label_b": label_b}}. Then, in `lit_module.py`, we index label from each model's output to compute the loss and metric, instead of using the label from `batch`. To make this design consistent across all models, we also need to modify the return of other models to include its label information.

File: text/src/autogluon/text/automm/data/mixup.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
+from timm.data.mixup import Mixup
+
+class MixupModule(Mixup):
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/data/mixup.py:3]: If we use mixup inside a model's forward, e.g., `timm_image.py`, we don't need this wrapper since the keys are not necessary.

File: text/src/autogluon/text/automm/optimization/lit_module.py [hunk 13]
```diff
@@ -174,7 +178,12 @@ def training_step(self, batch, batch_idx):
         -------
         Average loss of the mini-batch data.
         """
-        output, loss = self._shared_step(batch)
+        if self.mixup_fn is not None and (self.hparams.mixup_off_epoch is None or self.current_epoch < self.hparams.mixup_off_epoch):
+            batch, target_a, target_b, lam = self.mixup_fn(batch, "timm_image_image", self.model.label_key)
+            output, _ = self._shared_step(batch)
+            loss = self._compute_loss(output=output, label=target_a) * lam + self._compute_loss(output=output,label=target_b) * (1. - lam)
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/lit_module.py:184]: Why doesn't timm compute the loss in this way? https://github.com/rwightman/pytorch-image-models/blob/master/train.py#L703

Comment (Linuxdex) [reply text/src/autogluon/text/automm/optimization/lit_module.py:184]: The mixup in timm can only be used in specific loss function. https://github.com/rwightman/pytorch-image-models/blob/d30685c283137b4b91ea43c4e595c964cd2cb6f0/train.py#L587-L592
The target in the output is the one-hot code of the mixed label which is related with the specific number of classes. https://github.com/rwightman/pytorch-image-models/blob/d30685c283137b4b91ea43c4e595c964cd2cb6f0/timm/data/mixup.py#L25-L27

Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/lit_module.py:184]: Can we also follow timm's practice to use its customized loss function?

Comment (Linuxdex) [reply text/src/autogluon/text/automm/optimization/lit_module.py:184]: It is fine to use timm's loss functions in Multiclass and Binary. But it may face problems in Regression.

Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/lit_module.py:184]: We can define a similar loss class for regression. We choose the proper loss class based on the problem type. Using the modular design can make our training loop cleaner and scalable.

File: text/src/autogluon/text/automm/data/mixup.py (hunk 12 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/data/mixup.py:3]: I have the same question. Why do we need this class?

File: text/src/autogluon/text/automm/data/process_image.py [hunk 14]
```diff
@@ -269,17 +273,23 @@ def construct_processor(
             elif trans_mode == "vertical_flip":
                 processor.append(transforms.RandomVerticalFlip())
             elif trans_mode == "color_jitter":
-                if args is not None:
+                if kargs is not None:
+                    processor.append(transforms.ColorJitter(**kargs))
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/data/process_image.py:277]: We may document the usage

File: text/src/autogluon/text/automm/optimization/lit_module.py (hunk 13 above)
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/optimization/lit_module.py:184]: There is no need to define a new loss class for regression. We only need to inherit timm's Mixup class and add a new target mixup util function for regression and a flag to choose which target mixup.

File: text/src/autogluon/text/automm/utils.py [hunk 15]
```diff
@@ -1232,3 +1233,31 @@ def turn_on_off_feature_column_info(
                 per_model_processor.requires_column_info = flag
 
     return data_processors
+
+def get_mixup(
+        config: DictConfig,
+        output_shape: int,
+):
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/utils.py:1240]: Add document here. Also, refer to the mixup/cutmix paper and the implementation in TIMM in the documentation

**Comment by sxjscience** [comment]
@Linuxdex What's the status of this PR?

File: text/src/autogluon/text/automm/configs/data/default.yaml [hunk 16]
```diff
@@ -10,3 +10,12 @@ data:
     convert_to_text: False  # Whether to convert the feature to text.
     scaler_with_mean: True  # Whether to normalize with mean.
     scaler_with_std: True  # Whether to normalize with std.
+  mixup:
+    mixup: 0.8
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/configs/data/default.yaml:14]: We'd better also add documentation here.

File: text/src/autogluon/text/automm/data/process_image.py (hunk 10 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/data/process_image.py:295]: May be we can do it in a separate PR.

File: text/src/autogluon/text/automm/predictor.py [hunk 17]
```diff
@@ -412,7 +414,16 @@ def fit(
...
         )
 
-        loss_func = get_loss_func(problem_type)
+        mixup_active, mixup_fn = get_mixup(df_preprocessor=df_preprocessor,
+                                     config=config.data.mixup,
+                                     num_classes=output_shape,
+        )
+        if mixup_active and config.env.per_gpu_batch_size == 1:
+            warnings.warn("The mixup is done on the batch."
+                          "The per_gpu_batch_size should be >=1 for reasonable operation",
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/predictor.py:423]: per_gpu_batch_size should be >1

**Comment by sxjscience** [comment]
@Linuxdex Will you merge upstream/master and trigger CI again? The problem in core has been fixed.

File: text/src/autogluon/text/automm/optimization/soft_target_crossentropy.py [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+    It works under the mixup.
+    It can calculate the crossentropy of input and label with one-hot.
+    """
+
+    def __init__(self):
+        super(SoftTargetCrossEntropy, self).__init__()
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        loss = torch.sum(-target * F.log_softmax(input, dim=-1), dim=-1)
+        return loss.mean()
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/optimization/soft_target_crossentropy.py:19]: nit: new line

**Review by sxjscience — APPROVED** [review]
LGTM

File: text/src/autogluon/text/automm/configs/data/default.yaml [hunk 19]
```diff
@@ -11,3 +11,14 @@ data:
     scaler_with_mean: True  # Whether to normalize with mean.
     scaler_with_std: True  # Whether to normalize with std.
   pos_label:  # The name of binary classification's positive class. It's used in computing some metrics, e.g., roc_auc. If not provided, then use label_encoder.classes_[1],
+  mixup:
+    turn_on: True # The total control of mixup.
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/configs/data/default.yaml:15]: Should we change the default value to False?

**Comment by szha** [comment]
Job PR-1730-30 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1730/30/index.html

File: text/src/autogluon/text/automm/utils.py [hunk 20]
```diff
@@ -1261,3 +1261,58 @@ def try_to_infer_pos_label(
...
+    names = model_config.names
+    if isinstance(names, str):
+        names = [names]
+    for model_name in names:
+        permodel_config = getattr(model_config, model_name)
+        if hasattr(permodel_config.data_types, IMAGE):
+            model_active = True
+            break
+
+    mixup_active = False
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/utils.py:1298]: Why appearing twice?

**Comment by szha** [comment]
Job PR-1730-32 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1730/32/index.html
